cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+)
+
+var RolloutClock = clock.Clock(clock.RealClock{})
+var maxTimeDuration = time.Duration(math.MaxInt64)
+
+type RolloutStatus int
+
+const (
+	// resource desired status is not applied yet
+	ToApply RolloutStatus = iota
+	// resource desired status is applied and last applied status is not updated
+	Progressing
+	// resource desired status is applied and last applied status is succeed
+	Succeed
+	// resource desired status is applied and last applied status is failed
+	Failed
+	// when the rollout status is progressing or failed and the status remains for longer than the timeout
+	// then the status will be set to timeout
+	TimeOut
+	// skip rollout on this cluster
+	Skip
+)
+
+// Return the rollout status on each managed cluster
+type ClusterRolloutStatusFunc func(clusterName string) ClusterRolloutStatus
+
+type ClusterRolloutStatus struct {
+	// cluster name
+	// required field
+	Name string
+	// cluster group key
+	// optional field
+	GroupKey clusterv1beta1.GroupKey
+	// rollout status
+	// required field
+	Status RolloutStatus
+	// the last transition time of rollout status
+	// this is used to calculate timeout for progressing and failed status
+	// optional field
+	LastTransitionTime *time.Time
+}
+
+type RolloutResult struct {
+	// clusters to rollout
+	ClustersToRollout map[string]ClusterRolloutStatus
+	// clusters that are timeout
+	ClustersTimeOut map[string]ClusterRolloutStatus
+}
+
+// +k8s:deepcopy-gen=false
+type RolloutHandler struct {
+	// placement decision tracker
+	pdTracker *clusterv1beta1.PlacementDecisionClustersTracker
+}
+
+func NewRolloutHandler(pdTracker *clusterv1beta1.PlacementDecisionClustersTracker) *RolloutHandler {
+	if pdTracker == nil {
+		return nil
+	}
+
+	return &RolloutHandler{
+		pdTracker: pdTracker,
+	}
+}
+
+// The input is a duck type RolloutStrategy and a ClusterRolloutStatusFunc to return the rollout status on each managed cluster.
+// Return the strategy actual take effect and a list of cluster that needs to update with current state.
+func (r *RolloutHandler) GetRolloutCluster(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+	switch rolloutStrategy.Type {
+	case All:
+		return r.getRolloutAllClusters(rolloutStrategy, statusFunc)
+	case Progressive:
+		return r.getProgressiveClusters(rolloutStrategy, statusFunc)
+	case ProgressivePerGroup:
+		return r.getProgressivePerGroupClusters(rolloutStrategy, statusFunc)
+	default:
+		return nil, RolloutResult{}, fmt.Errorf("incorrect rollout strategy type %v", rolloutStrategy.Type)
+	}
+}
+
+func (r *RolloutHandler) getRolloutAllClusters(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+	strategy := RolloutStrategy{Type: All}
+	strategy.All = rolloutStrategy.All.DeepCopy()
+	if strategy.All == nil {
+		strategy.All = &RolloutAll{}
+	}
+
+	// parse timeout
+	failureTimeout, err := parseTimeout(strategy.All.Timeout.Timeout)
```

Comment:
What does a timeout do for the all strategy? In the policy example, presumably all clusters would be assigned the policy and if they are connected to the hub, they would copy the policy down and evaluate it.

Reply:
For the all strategy, all the clusters would apply the new resource at once, some might keep in progressing or failed status, and if there's a timeout, the controller won't keep waiting for it to reach a successful state. 
cc @serngawy , correct me if not accurate.

Reply:
yes, for some consumer controller (eg; TALM) even with ALL rollout it needs to set a state as Successful or Failed after Timeout (It will not continue to reconcile all the times, there is a maintenance window time constrain)